consensus/src/dag/reliable_broadcast.rs

```diff
@@ -67,3 +80,118 @@ impl ReliableBroadcast {
         }
     }
 }
+
+#[derive(ThisError, Debug)]
+pub enum NodeBroadcastHandleError {
+    #[error("invalid parent in node")]
+    InvalidParent,
+    #[error("missing parents")]
+    MissingParents,
+    #[error("parents do not meet quorum voting power")]
+    NotEnoughParents,
+}
+
+pub struct NodeBroadcastHandler {
+    dag: Arc<RwLock<Dag>>,
+    signatures_by_round_peer: BTreeMap<Round, BTreeMap<Author, NodeDigestSignature>>,
+    signer: ValidatorSigner,
+    verifier: ValidatorVerifier,
+}
+
+impl NodeBroadcastHandler {
+    pub fn new(
+        dag: Arc<RwLock<Dag>>,
+        signer: ValidatorSigner,
+        verifier: ValidatorVerifier,
+    ) -> Self {
+        Self {
+            dag,
+            signatures_by_round_peer: BTreeMap::new(),
+            signer,
+            verifier,
+        }
+    }
+
+    pub fn gc_before_round(&mut self, min_round: Round) {
+        self.signatures_by_round_peer.retain(|r, _| r >= &min_round);
+    }
+
+    fn validate(&self, node: &Node) -> anyhow::Result<()> {
+        let current_round = node.metadata().round();
+        // round 0 is a special case and does not require any parents
+        if current_round == 0 {
+            return Ok(());
+        }
+
+        node.verify(&self.verifier)?;
+
+        let prev_round = current_round - 1;
+
+        let dag_reader = self.dag.read();
+        // check if the parent round is missing in the DAG
+        ensure!(
+            prev_round >= dag_reader.lowest_round(),
+            NodeBroadcastHandleError::MissingParents
+        );
+
+        // check which parents are missing in the DAG
+        let missing_parents: Vec<NodeCertificate> = node
+            .parents()
+            .iter()
+            .filter(|parent| !dag_reader.exists(parent.metadata().digest()))
+            .cloned()
+            .collect();
+        if !missing_parents.is_empty() {
+            // For each missing parent, verify their signatures and voting power
+            ensure!(
+                missing_parents.iter().all(|parent| {
+                    let node_digest = NodeDigest::new(*parent.metadata().digest());
+                    self.verifier
+                        .verify_multi_signatures(&node_digest, parent.signatures())
```

Comment:
I think this checks voting power inside, I'd slightly prefer we do all the self-contained checks (including parents certificate check) for node outside but let's do that in another pr.